File: ios/RealCord/Models/Message.swift
import Foundation
import FirebaseFirestore

struct Message: Codable, Identifiable, Equatable {
    @DocumentID var id: String?
    var channelId: String
    var senderId: String
    var senderName: String
    var senderAvatar: String?
    var content: String
    var imageUrl: String?
    var fileUrl: String?
    var fileName: String?
    var createdAt: Timestamp?
    var editedAt: Timestamp?
    var replyToId: String?
    var mentions: [String]?
    var reactions: [String: Int]?
    var userReactions: [String: String]?
    var readAt: Timestamp?

    enum CodingKeys: String, CodingKey {
        case channelId, senderId, senderName, senderAvatar, content
        case imageUrl, fileUrl, fileName, createdAt, editedAt
        case replyToId, mentions, reactions, userReactions, readAt
    }

    var isSent: Bool = false
    var isRead: Bool = false

    mutating func setLocalFlags(isSent: Bool, isRead: Bool) {
        self.isSent = isSent
        self.isRead = isRead
    }
}

File: android/app/src/main/java/com/discordclone/models/Message.java
package com.discordclone.models;

import com.google.firebase.Timestamp;
import java.util.List;
import java.util.Map;

public class Message {
    private String id;
    private String channelId;
    private String senderId;
    private String senderName;
    private String senderAvatar;
    private String content;
    private String imageUrl;
    private String fileUrl;
    private String fileName;
    private Timestamp createdAt;
    private Timestamp editedAt;
    private Map<String, Integer> reactions;
    private String replyToId;
    private List<String> mentions;
    private Map<String, String> userReactions;
    private Timestamp readAt;

    public Message() {}

    public Message(String channelId, String senderId, String senderName, String content) {
        this.channelId = channelId;
        this.senderId = senderId;
        this.senderName = senderName;
        this.content = content;
        this.createdAt = Timestamp.now();
    }

    public String getId() { return id; }
    public void setId(String id) { this.id = id; }

    public String getChannelId() { return channelId; }
    public void setChannelId(String channelId) { this.channelId = channelId; }

    public String getSenderId() { return senderId; }
    public void setSenderId(String senderId) { this.senderId = senderId; }

    public String getSenderName() { return senderName; }
    public void setSenderName(String senderName) { this.senderName = senderName; }

    public String getSenderAvatar() { return senderAvatar; }
    public void setSenderAvatar(String senderAvatar) { this.senderAvatar = senderAvatar; }

    public String getContent() { return content; }
    public void setContent(String content) { this.content = content; }

    public String getImageUrl() { return imageUrl; }
    public void setImageUrl(String imageUrl) { this.imageUrl = imageUrl; }

    public String getFileUrl() { return fileUrl; }
    public void setFileUrl(String fileUrl) { this.fileUrl = fileUrl; }

    public String getFileName() { return fileName; }
    public void setFileName(String fileName) { this.fileName = fileName; }

    public Timestamp getCreatedAt() { return createdAt; }
    public void setCreatedAt(Timestamp createdAt) { this.createdAt = createdAt; }

    public Timestamp getEditedAt() { return editedAt; }
    public void setEditedAt(Timestamp editedAt) { this.editedAt = editedAt; }

    public Map<String, Integer> getReactions() { return reactions; }
    public void setReactions(Map<String, Integer> reactions) { this.reactions = reactions; }

    public String getReplyToId() { return replyToId; }
    public void setReplyToId(String replyToId) { this.replyToId = replyToId; }

    public List<String> getMentions() { return mentions; }
    public void setMentions(List<String> mentions) { this.mentions = mentions; }

    public Map<String, String> getUserReactions() { return userReactions; }
    public void setUserReactions(Map<String, String> userReactions) { this.userReactions = userReactions; }

    public Timestamp getReadAt() { return readAt; }
    public void setReadAt(Timestamp readAt) { this.readAt = readAt; }
}
